Speed up `fmpz_gcd`

albinahlback opened the pull request:
With the profiler code, I get
```
  5 bits:   (min) 0.97x speedup   (max) 0.93x speedup
 10 bits:   (min) 1.05x speedup   (max) 1.05x speedup
 15 bits:   (min) 1.07x speedup   (max) 1.06x speedup
 20 bits:   (min) 1.09x speedup   (max) 1.08x speedup
 25 bits:   (min) 1.11x speedup   (max) 1.11x speedup
 30 bits:   (min) 1.12x speedup   (max) 1.12x speedup
 35 bits:   (min) 1.14x speedup   (max) 1.14x speedup
 40 bits:   (min) 1.15x speedup   (max) 1.15x speedup
 45 bits:   (min) 1.16x speedup   (max) 1.16x speedup
 50 bits:   (min) 1.18x speedup   (max) 1.18x speedup
 55 bits:   (min) 1.19x speedup   (max) 1.19x speedup
 60 bits:   (min) 1.20x speedup   (max) 1.19x speedup
 65 bits:   (min) 1.20x speedup   (max) 1.20x speedup
 70 bits:   (min) 1.19x speedup   (max) 1.19x speedup
 75 bits:   (min) 1.18x speedup   (max) 1.19x speedup
 80 bits:   (min) 1.17x speedup   (max) 1.16x speedup
 85 bits:   (min) 1.16x speedup   (max) 1.16x speedup
 90 bits:   (min) 1.15x speedup   (max) 1.13x speedup
 95 bits:   (min) 1.14x speedup   (max) 1.12x speedup
100 bits:   (min) 1.14x speedup   (max) 1.14x speedup
105 bits:   (min) 1.12x speedup   (max) 1.12x speedup
110 bits:   (min) 1.12x speedup   (max) 1.04x speedup
115 bits:   (min) 1.11x speedup   (max) 1.09x speedup
120 bits:   (min) 1.11x speedup   (max) 1.10x speedup
125 bits:   (min) 1.10x speedup   (max) 1.08x speedup
130 bits:   (min) 1.10x speedup   (max) 1.09x speedup
135 bits:   (min) 1.09x speedup   (max) 1.09x speedup
140 bits:   (min) 1.09x speedup   (max) 1.09x speedup
145 bits:   (min) 1.09x speedup   (max) 1.10x speedup
150 bits:   (min) 1.08x speedup   (max) 1.07x speedup
```

Edit: This was made on an i5 6300U.

albinahlback commented:
Why does it say undefined reference to `mpn_gcd_11`?

tthsqe12 commented:
because the actual name is _gmpn_gcd_11 or something like that?
n_gcd can replaced with mpnz_gcd_11 on some setups, but, @wbhart, are we allowed to do this in general?

fredrik-johansson commented:
Alternatively, can we make n_gcd as fast as mpn_gcd_11?

tthsqe12 commented:
No, x86 compilers cannot/refuse to output the sequence of instructions for the intel flavors of mpn_gcd_11.

albinahlback commented:
> because the actual name is _gmpn_gcd_11 or something like that? n_gcd can replaced with mpnz_gcd_11 on some setups, but, @wbhart, are we allowed to do this in general?

No, the function is called `mpn_gcd_11`...

tthsqe12 commented:
... it could be a macro :). So, whatever `gmp.h` the test machines are using, it is clear that it doesn't have this line
```
#define mpn_gcd_11 __MPN(gcd_11)
__GMP_DECLSPEC mp_limb_t mpn_gcd_11 (mp_limb_t, mp_limb_t) __GMP_ATTRIBUTE_PURE;
```
This is another argument for turning this implicit declarations into errors.

albinahlback commented:
Would you know why it works on my computer but not on others? I mean, I think GMP has had that interface for quite a while, so it shouldn't be that my version is too new?

tthsqe12 commented:
Your bax is finding a `gmp.h` with that line in it, and your linker is finding a `libgmp` with the symbol `__gmpn_gcd_11`. This is not true of whatever the CI is doing.

albinahlback commented:
It couldn't find `__gmpn_gcd_11` either. I will just link it to `n_gcd` if it can't be found.

wbhart commented:
mpn_gcd_11 is not in the GMP interface. I just searched the documentation for the latest GMP and it has either been taken out or was never there. This means we cannot link to it reliably, as they can remove it at any time of their choosing (and we should assume that they will). They could even remove the function and not the macro, and then this code could fail.

I simply would not do this. I think Fredrik's answer is the better one, though as usual we cannot take the code from GMP itself. We have to figure out ourselves how it could be done and then do it.

I am honestly quite surprised that GMP's version is faster than ours given how many times people have replaced it with a faster version.

albinahlback commented:
If you download GMP, you can see the version used under `mpn/x86_64/core2/gcd_11.asm` I believe. To simply put it, it is super fast under the assumption that the inputs are odd.

Perhaps one could mimic it, but I believe the current `mpn_gcd_11` will outperform everything.

tthsqe12 commented:
@albinahlback you can try to speed up n_gcd on x86 by writing branchless code inside the loop. Last time I tried it, it got fairly close (20% maybe)  to mpngcd11, but I doubt you can get much closer in C.

albinahlback commented:
Yeah, sure I can try. That will be some good training for me at least.

wbhart commented:
Yeah I see they have it in assembly now. Makes sense of course. Oh well, we can try Dan's suggestion. But short of writing one hell of a lot of assembly versions for different processors, I think we are stuck without this function unless they decide to make it an official part of the interface.

Probably calling mpn_gcd_1 is worth experimenting with too (that is in the official interface). That should call mpn_gcd_11 in the trivial case, so should only be some extra function call overhead and a branch, both of which are not very expensive on modern CPUs.

albinahlback commented:
I cannot see that it anywhere says that `mpn_gcd_11` is not intended for users of GMP. I mean, it is even in `gmp.h` under the same section as `mpn_gcd_1`.

I asked a question about it on https://gmplib.org/list-archives/gmp-discuss/2021-November/thread.html, but I think it is very reasonable that it actually is intended for users as well, just that it is not documented.

wbhart commented:
As you see, the maintainer of GMP has now responded and says that you should use the documented mpn_gcd_1 instead.

albinahlback commented:
Yes, I saw that! I shall now try to complete this PR.

albinahlback commented:
Okay, so I think I am satisfied with the current state (my latest commit). The results are
```
  5 bits:   (min) 1.00x speedup   (max) 1.00x speedup
 10 bits:   (min) 1.03x speedup   (max) 1.02x speedup
 15 bits:   (min) 1.05x speedup   (max) 1.05x speedup
 20 bits:   (min) 1.06x speedup   (max) 1.07x speedup
 25 bits:   (min) 1.08x speedup   (max) 1.07x speedup
 30 bits:   (min) 1.09x speedup   (max) 1.09x speedup
 35 bits:   (min) 1.10x speedup   (max) 1.11x speedup
 40 bits:   (min) 1.12x speedup   (max) 1.11x speedup
 45 bits:   (min) 1.13x speedup   (max) 1.13x speedup
 50 bits:   (min) 1.14x speedup   (max) 1.14x speedup
 55 bits:   (min) 1.15x speedup   (max) 1.15x speedup
 60 bits:   (min) 1.16x speedup   (max) 1.17x speedup
 65 bits:   (min) 1.17x speedup   (max) 1.16x speedup
 70 bits:   (min) 1.15x speedup   (max) 1.15x speedup
 75 bits:   (min) 1.14x speedup   (max) 1.14x speedup
 80 bits:   (min) 1.13x speedup   (max) 1.13x speedup
 85 bits:   (min) 1.12x speedup   (max) 1.12x speedup
 90 bits:   (min) 1.12x speedup   (max) 1.11x speedup
 95 bits:   (min) 1.11x speedup   (max) 1.13x speedup
100 bits:   (min) 1.11x speedup   (max) 1.11x speedup
105 bits:   (min) 1.10x speedup   (max) 1.10x speedup
110 bits:   (min) 1.10x speedup   (max) 1.09x speedup
115 bits:   (min) 1.09x speedup   (max) 1.09x speedup
120 bits:   (min) 1.09x speedup   (max) 1.09x speedup
125 bits:   (min) 1.08x speedup   (max) 1.08x speedup
130 bits:   (min) 1.08x speedup   (max) 1.08x speedup
135 bits:   (min) 1.07x speedup   (max) 1.07x speedup
140 bits:   (min) 1.07x speedup   (max) 1.07x speedup
145 bits:   (min) 1.07x speedup   (max) 1.08x speedup
150 bits:   (min) 1.06x speedup   (max) 1.06x speedup
```
which is quite nice. I tried replacing `mpz_gcd` with `mpn_gcd`, but I could not make it faster.

wbhart commented:
Again, is there any reason to have the new code in the profiling code rather than to just call the new code? One of the problems with having any code in the same file is that different optimisations can be applied by the compiler that couldn't be applied if it had to be called in the library.

albinahlback commented:
That's true. I now removed that and squashed all commits (hence the force-push). These are the final timings on Haswell 4770k:
```
  5 bits:   (min) 0.96x speedup   (max) 0.96x speedup
 10 bits:   (min) 0.98x speedup   (max) 0.98x speedup
 15 bits:   (min) 1.00x speedup   (max) 1.00x speedup
 20 bits:   (min) 1.02x speedup   (max) 1.02x speedup
 25 bits:   (min) 1.04x speedup   (max) 1.04x speedup
 30 bits:   (min) 1.06x speedup   (max) 1.03x speedup
 35 bits:   (min) 1.07x speedup   (max) 1.07x speedup
 40 bits:   (min) 1.09x speedup   (max) 1.09x speedup
 45 bits:   (min) 1.11x speedup   (max) 1.10x speedup
 50 bits:   (min) 1.13x speedup   (max) 1.13x speedup
 55 bits:   (min) 1.15x speedup   (max) 1.15x speedup
 60 bits:   (min) 1.17x speedup   (max) 1.16x speedup
 65 bits:   (min) 1.17x speedup   (max) 1.17x speedup
 70 bits:   (min) 1.16x speedup   (max) 1.15x speedup
 75 bits:   (min) 1.15x speedup   (max) 1.15x speedup
 80 bits:   (min) 1.14x speedup   (max) 1.14x speedup
 85 bits:   (min) 1.13x speedup   (max) 1.13x speedup
 90 bits:   (min) 1.13x speedup   (max) 1.12x speedup
 95 bits:   (min) 1.12x speedup   (max) 1.12x speedup
100 bits:   (min) 1.12x speedup   (max) 1.11x speedup
105 bits:   (min) 1.11x speedup   (max) 1.11x speedup
110 bits:   (min) 1.11x speedup   (max) 1.10x speedup
115 bits:   (min) 1.10x speedup   (max) 1.10x speedup
120 bits:   (min) 1.10x speedup   (max) 0.99x speedup
125 bits:   (min) 1.09x speedup   (max) 1.00x speedup
130 bits:   (min) 1.09x speedup   (max) 1.07x speedup
135 bits:   (min) 1.08x speedup   (max) 1.07x speedup
140 bits:   (min) 1.08x speedup   (max) 1.06x speedup
145 bits:   (min) 1.08x speedup   (max) 1.08x speedup
150 bits:   (min) 1.07x speedup   (max) 1.07x speedup
```

wbhart commented:
That looks convincing. I'll merge when appveyor has finished.

fredrik-johansson commented:
Might also want to use mpn_gcd_1 in fmpz_gcd3.

By the way, if it's always faster, shouldn't we redefine n_gcd to wrap mpn_gcd_1?

albinahlback commented:
> Might also want to use mpn_gcd_1 in fmpz_gcd3.

Yes, I will now try to speed up `fmpz_gcd3` as well.

> By the way, if it's always faster, shouldn't we redefine n_gcd to wrap mpn_gcd_1?

Sounds good to me. Then I also think it's a good idea to introduce `fmpz_gcd_ui` which also just wraps `mpn_gcd_1`.

albinahlback commented:
New version of `fmpz_gcd3`:
```
  1 bits:   (min) 0.93x speedup   (max) 0.93x speedup
  4 bits:   (min) 0.94x speedup   (max) 0.93x speedup
  7 bits:   (min) 0.94x speedup   (max) 0.94x speedup
 10 bits:   (min) 0.96x speedup   (max) 0.95x speedup
 13 bits:   (min) 0.97x speedup   (max) 0.96x speedup
 16 bits:   (min) 0.98x speedup   (max) 0.98x speedup
 19 bits:   (min) 0.99x speedup   (max) 0.98x speedup
 22 bits:   (min) 1.00x speedup   (max) 0.99x speedup
 25 bits:   (min) 1.00x speedup   (max) 1.00x speedup
 28 bits:   (min) 1.01x speedup   (max) 1.01x speedup
 31 bits:   (min) 1.02x speedup   (max) 1.02x speedup
 34 bits:   (min) 1.03x speedup   (max) 1.03x speedup
 37 bits:   (min) 1.04x speedup   (max) 1.03x speedup
 40 bits:   (min) 1.05x speedup   (max) 1.06x speedup
 43 bits:   (min) 1.06x speedup   (max) 1.05x speedup
 46 bits:   (min) 1.07x speedup   (max) 1.06x speedup
 49 bits:   (min) 1.08x speedup   (max) 1.08x speedup
 52 bits:   (min) 1.08x speedup   (max) 1.08x speedup
 55 bits:   (min) 1.10x speedup   (max) 1.09x speedup
 58 bits:   (min) 1.10x speedup   (max) 1.10x speedup
 61 bits:   (min) 1.11x speedup   (max) 1.11x speedup
 64 bits:   (min) 1.11x speedup   (max) 1.11x speedup
 67 bits:   (min) 1.09x speedup   (max) 1.08x speedup
 70 bits:   (min) 1.08x speedup   (max) 1.08x speedup
 73 bits:   (min) 1.07x speedup   (max) 1.07x speedup
 76 bits:   (min) 1.07x speedup   (max) 1.06x speedup
 79 bits:   (min) 1.06x speedup   (max) 1.05x speedup
 82 bits:   (min) 1.07x speedup   (max) 1.06x speedup
 85 bits:   (min) 1.06x speedup   (max) 1.06x speedup
 88 bits:   (min) 1.06x speedup   (max) 1.05x speedup
 91 bits:   (min) 1.06x speedup   (max) 1.06x speedup
 94 bits:   (min) 1.06x speedup   (max) 1.06x speedup
 97 bits:   (min) 1.06x speedup   (max) 1.06x speedup
100 bits:   (min) 1.06x speedup   (max) 1.06x speedup
103 bits:   (min) 1.06x speedup   (max) 1.06x speedup
106 bits:   (min) 1.06x speedup   (max) 1.06x speedup
109 bits:   (min) 1.06x speedup   (max) 1.03x speedup
112 bits:   (min) 1.05x speedup   (max) 1.03x speedup
115 bits:   (min) 1.05x speedup   (max) 1.05x speedup
118 bits:   (min) 1.05x speedup   (max) 1.05x speedup
121 bits:   (min) 1.05x speedup   (max) 1.05x speedup
124 bits:   (min) 1.05x speedup   (max) 1.03x speedup
127 bits:   (min) 1.05x speedup   (max) 1.05x speedup
130 bits:   (min) 1.05x speedup   (max) 1.04x speedup
133 bits:   (min) 1.05x speedup   (max) 1.05x speedup
136 bits:   (min) 1.05x speedup   (max) 1.05x speedup
139 bits:   (min) 1.05x speedup   (max) 1.04x speedup
142 bits:   (min) 1.05x speedup   (max) 1.05x speedup
145 bits:   (min) 1.05x speedup   (max) 1.05x speedup
148 bits:   (min) 1.05x speedup   (max) 1.04x speedup
151 bits:   (min) 1.05x speedup   (max) 1.05x speedup
154 bits:   (min) 1.05x speedup   (max) 1.05x speedup
157 bits:   (min) 1.05x speedup   (max) 1.05x speedup
160 bits:   (min) 1.05x speedup   (max) 1.05x speedup
163 bits:   (min) 1.05x speedup   (max) 1.04x speedup
166 bits:   (min) 1.05x speedup   (max) 1.05x speedup
169 bits:   (min) 1.05x speedup   (max) 1.04x speedup
172 bits:   (min) 1.05x speedup   (max) 1.04x speedup
175 bits:   (min) 1.04x speedup   (max) 1.04x speedup
178 bits:   (min) 1.05x speedup   (max) 1.02x speedup
181 bits:   (min) 1.04x speedup   (max) 1.02x speedup
184 bits:   (min) 1.04x speedup   (max) 1.06x speedup
187 bits:   (min) 1.04x speedup   (max) 1.04x speedup
190 bits:   (min) 1.04x speedup   (max) 1.04x speedup
193 bits:   (min) 1.04x speedup   (max) 1.04x speedup
196 bits:   (min) 1.04x speedup   (max) 1.04x speedup
199 bits:   (min) 1.04x speedup   (max) 1.02x speedup
```
which can be compared to the old version in the benchmark (just to make sure that it actually is faster in all those cases):
```
  1 bits:   (min) 0.92x speedup   (max) 0.94x speedup
  4 bits:   (min) 0.93x speedup   (max) 0.93x speedup
  7 bits:   (min) 0.93x speedup   (max) 0.93x speedup
 10 bits:   (min) 0.94x speedup   (max) 0.94x speedup
 13 bits:   (min) 0.94x speedup   (max) 0.93x speedup
 16 bits:   (min) 0.94x speedup   (max) 0.94x speedup
 19 bits:   (min) 0.94x speedup   (max) 0.94x speedup
 22 bits:   (min) 0.94x speedup   (max) 0.94x speedup
 25 bits:   (min) 0.94x speedup   (max) 0.94x speedup
 28 bits:   (min) 0.95x speedup   (max) 0.94x speedup
 31 bits:   (min) 0.94x speedup   (max) 0.94x speedup
 34 bits:   (min) 0.95x speedup   (max) 0.94x speedup
 37 bits:   (min) 0.94x speedup   (max) 0.94x speedup
 40 bits:   (min) 0.95x speedup   (max) 0.95x speedup
 43 bits:   (min) 0.95x speedup   (max) 0.95x speedup
 46 bits:   (min) 0.95x speedup   (max) 0.95x speedup
 49 bits:   (min) 0.95x speedup   (max) 0.95x speedup
 52 bits:   (min) 0.95x speedup   (max) 0.95x speedup
 55 bits:   (min) 0.95x speedup   (max) 0.95x speedup
 58 bits:   (min) 0.95x speedup   (max) 0.95x speedup
 61 bits:   (min) 0.95x speedup   (max) 0.94x speedup
 64 bits:   (min) 0.95x speedup   (max) 0.95x speedup
 67 bits:   (min) 0.96x speedup   (max) 0.96x speedup
 70 bits:   (min) 0.96x speedup   (max) 0.96x speedup
 73 bits:   (min) 0.97x speedup   (max) 0.97x speedup
 76 bits:   (min) 0.97x speedup   (max) 0.98x speedup
 79 bits:   (min) 0.97x speedup   (max) 0.97x speedup
 82 bits:   (min) 0.97x speedup   (max) 0.97x speedup
 85 bits:   (min) 0.97x speedup   (max) 0.97x speedup
 88 bits:   (min) 0.97x speedup   (max) 0.97x speedup
 91 bits:   (min) 0.98x speedup   (max) 0.97x speedup
 94 bits:   (min) 0.97x speedup   (max) 0.97x speedup
 97 bits:   (min) 0.98x speedup   (max) 0.98x speedup
100 bits:   (min) 0.98x speedup   (max) 0.97x speedup
103 bits:   (min) 0.98x speedup   (max) 0.97x speedup
106 bits:   (min) 0.98x speedup   (max) 0.96x speedup
109 bits:   (min) 0.98x speedup   (max) 0.98x speedup
112 bits:   (min) 0.98x speedup   (max) 0.98x speedup
115 bits:   (min) 0.98x speedup   (max) 0.98x speedup
118 bits:   (min) 0.98x speedup   (max) 0.98x speedup
121 bits:   (min) 0.98x speedup   (max) 0.97x speedup
124 bits:   (min) 0.98x speedup   (max) 0.98x speedup
127 bits:   (min) 0.98x speedup   (max) 0.98x speedup
130 bits:   (min) 0.98x speedup   (max) 0.98x speedup
133 bits:   (min) 0.98x speedup   (max) 0.98x speedup
136 bits:   (min) 0.98x speedup   (max) 0.98x speedup
139 bits:   (min) 0.98x speedup   (max) 0.98x speedup
142 bits:   (min) 0.98x speedup   (max) 0.98x speedup
145 bits:   (min) 0.98x speedup   (max) 0.98x speedup
148 bits:   (min) 0.98x speedup   (max) 0.98x speedup
151 bits:   (min) 0.98x speedup   (max) 0.96x speedup
154 bits:   (min) 0.98x speedup   (max) 0.98x speedup
157 bits:   (min) 0.98x speedup   (max) 0.97x speedup
160 bits:   (min) 0.98x speedup   (max) 0.98x speedup
163 bits:   (min) 0.98x speedup   (max) 0.98x speedup
166 bits:   (min) 0.98x speedup   (max) 0.97x speedup
169 bits:   (min) 0.98x speedup   (max) 0.98x speedup
172 bits:   (min) 0.98x speedup   (max) 0.98x speedup
175 bits:   (min) 0.98x speedup   (max) 0.98x speedup
178 bits:   (min) 0.98x speedup   (max) 0.98x speedup
181 bits:   (min) 0.98x speedup   (max) 0.98x speedup
184 bits:   (min) 0.98x speedup   (max) 0.98x speedup
187 bits:   (min) 0.96x speedup   (max) 0.91x speedup
190 bits:   (min) 0.98x speedup   (max) 0.98x speedup
193 bits:   (min) 0.98x speedup   (max) 0.98x speedup
196 bits:   (min) 0.98x speedup   (max) 0.98x speedup
199 bits:   (min) 0.98x speedup   (max) 0.97x speedup
```

albinahlback commented:
Ugh, didnt check aliasing... Is that crucial though?

albinahlback commented on ulong_extras/gcd.c:
```diff
@@ -1,6 +1,5 @@
 /*
-    Copyright (C) 2009, 2015 William Hart
-    Copyright (C) 2014 Abhinav Baid
+    Copyright (C) 2021 Albin Ahlbäck
```
Since the function got a complete makeover, I thought it was best to remove the previous names. Let me know if I should change it back.

wbhart commented:
Alias testing should be done for all user facing Flint functions.

albinahlback commented:
Yes I think I fixed it. I think this is ready for review now.

wbhart commented on fmpz/gcd.c:
```diff
@@ -11,61 +12,73 @@
-            __mpz_struct *mpz_ptr = _fmpz_promote(f);   /* aliasing fine as g, h already large */
-
+            /* TODO: Change to mpn_gcd in order to save some calculations that
+             * has already been already made. */
```
has -> have

albinahlback replied:
Thanks!

wbhart commented:
Looks good. I'll merge when you fix the trivial typo I pointed out.

fredrik-johansson commented:
There is a potential performance regression in gcd3. It is supposed to do the gcd of the smaller operands first but in the new _fmpz_gcd3_small you do an nx1 gcd before the 1x1 gcd when `a` is large and `b` is small.

albinahlback commented:
Hopefully this passes all tests. I also added some checks for `c  != 1` so we don't call `mpn_gcd_1` unless it is necessary.
